Blish HUD/Controls/TextInputBase.cs

```diff
@@ -0,0 +1,718 @@
+                this.SelectionEnd = _cursorIndex = _selectionStart;
+            } else {
+                Delete(_selectionEnd, _selectionStart - _selectionEnd);
+                this.SelectionStart = _cursorIndex = _selectionEnd;
+            }
+        }
+
+        private bool Paste(string value) {
+            DeleteSelection();
+
+            if (InsertChars(_cursorIndex, value, out var length) && length > 0) {
+                _undoStack.MakeInsert(_cursorIndex, length);
+                this.CursorIndex += length;
+                return true;
+            }
+
+            return false;
+        }
+
+        private void InputChar(char value) {
+            if (value == NEWLINE) {
+                if (!_multiline) return;
+            } else if (_font.GetCharacterRegion(value) == null) return;
+
+            if (_insertMode && _selectionStart == _selectionEnd && _cursorIndex < _text.Length) {
+                _undoStack.MakeReplace(_text, _cursorIndex, 1, 1);
+                DeleteChars(_cursorIndex, 1);
+
+                if (InsertChar(_cursorIndex, value)) {
+                    UserSetCursorIndex(_cursorIndex + 1);
+                }
+            } else {
+                DeleteSelection();
+
+                if (InsertChar(_cursorIndex, value)) {
+                    _undoStack.MakeInsert(_cursorIndex, 1);
+                    UserSetCursorIndex(_cursorIndex + 1);
+                }
+            }
+
+            ResetSelection();
+        }
+
+        private void UndoRedo(UndoRedoStack undoStack, UndoRedoStack redoStack) {
+            if (undoStack.Stack.Count == 0) return;
+
+            var record = undoStack.Stack.Pop();
+
+            try {
+                _suppressRedoStackReset = true;
+
+                switch (record.OperationType) {
+                    case OperationType.Insert:
+                        redoStack.MakeDelete(_text, record.Index, record.Length);
+                        DeleteChars(record.Index, record.Length);
+                        UserSetCursorIndex(record.Index);
+                        break;
+                    case OperationType.Delete:
+                        if (InsertChars(record.Index, record.Data, out int length)) {
+                            redoStack.MakeInsert(record.Index, length);
+                            UserSetCursorIndex(record.Index + length);
+                        }
+
+                        break;
+                    case OperationType.Replace:
+                        redoStack.MakeReplace(_text, record.Index, record.Length, record.Data.Length);
+                        DeleteChars(record.Index, record.Length);
+                        InsertChars(record.Index, record.Data, out _);
+                        break;
+                }
+            } finally {
+                _suppressRedoStackReset = false;
+            }
+
+            ResetSelection();
+        }
+
+        protected void UserSetCursorIndex(int newIndex) {
```

Comment:
Exactly for that is a Setter of a property and i don't see a reason why it shouldn't be there also for the internal setters in this class. I mean this check should also be done if you are inserting characters or smth like that. The only benefit would be that it will be faster but that's a pretty non existing performance boost (premature optimization is the root of all evil and so on)

Reply:
This is again to match the original implementation.

https://github.com/rds1983/Myra/blob/a9dbf7a1ceedc19f9e416c754eaf38e89a89a746/src/Myra/Graphics2D/UI/TextBox.cs#L545-L558

It's not for an optimization.

Reply:
Can be resolved